_posts/2020-09-27-spring-boot-12-factor-app.md

```diff
@@ -0,0 +1,192 @@
+
+> Maximize robustness with fast startup and graceful shutdown.
+
+Spring Boot applications are commonly run inside containers. Containers are ephemeral and can be started or stopped at any moment. 
+
+So it is important to minimize the startup time and ensure that the application shuts down gracefully when the container stops. Startup time is minimized with lazy initialization of dependent resources and by building [optimized container images](https://reflectoring.io/spring-boot-docker/).
+
+## Dev/Prod Parity - Build Once - Ship Anywhere
+> Keep development, staging, and production as similar as possible.
+
+Movement of code across environments has traditionally been a major factor slowing down the development velocity. This resulted from a difference in the infrastructure used for development and production. 
+
+Containers made this possible to build once and ship to multiple target environments. Containers make it possible to package all the dependencies including the OS and all dependencies. 
+
+Spring Boot applications are packaged in Docker containers and pushed to a Docker registry. Apart from using a Docker file to create a Docker image, Spring Boot provides plugins for [building OCI image from source](https://reflectoring.io/spring-boot-docker/) with Cloud-Native buildpacks.
+
+
+## Logs - Publish Logs as Event Streams
+> Treat Logs as Event Streams.
+
+The application should only produce logs in the form of event streams. Storing the logs in files or databases and shipping to other systems for further analysis should be delegated to purpose-built software. 
+
+Spring Boot logs only to the console by default and does not write log files. It is preconfigured with Logback as the default Logger implementation. However, the principle of producing logs as event streams enables it to be integrated with a rich ecosystem of log appenders, filters, shippers, monitoring, and visualization tools to build a highly observable system. All these are elaborated in [configuring logging in Spring boot](https://reflectoring.io/springboot-logging/).
+
+
+## Admin Processes - Built as API and Packaged with the Application
+> Run admin/management tasks as one-off processes.
+
+Most applications need to run one-off tasks for administration and management. Examples of these tasks include database scripts to initialize the database or scripts for fixing bad records. This code should be packaged with the application and released together and run in the same environment. 
+
+In a Spring Boot application, we expose admin functions as separate endpoints that are invoked as one-off processes. Adding functions to execute one-off processes will go through the build, test, and release cycle.
+
+
+## Conclusion
+We looked at the Twelve-factor principles for building a cloud-native application with Spring Boot. The following table summarizes what we have to do and what Spring Boot does for us to follow the 12 factors:
+
+| Factor        | What do we have to do?|
```

Comment:
I changed the title of the column to "What do we have to do?". I think that makes the table more actionable. Please go through the rows and change the entries so that they describe an action. 

I.e. "One codebase for all environments" becomes "Use one codebase for all environments". 

The first "Container workflow" becomes "Build and publish a Docker image"

and so on